app/models/discussion_reader.rb

```diff
@@ -97,12 +97,7 @@ def viewed!(age_of_last_read_item = nil)
       reset_counts!
     end
 
-    publish!
-  end
-
-  def publish!
-    publish_data = Simple::DiscussionSerializer.new(discussion, scope: { reader_cache: reader_cache })
-    MessageChannelService.publish(publish_data.as_json, to: user)
+    EventBus.broadcast('discussion_reader_viewed', discussion, self, user)
```

Comment:
Nah, but we can listen for both 'discussion_create' and
'discussion_mark_as_read' in the EventBus listener

```
EventBus.listen('discussion_create', 'discussion_mark_as_read') do
|discussion, actor|
  reader = get_the_reader(discussion, actor)
  serialize_the_thing(reader)
end
```

On Thu, Feb 18, 2016 at 8:51 AM, Robert Guthrie <notifications@github.com>
wrote:

> In app/models/discussion_reader.rb
> <https://github.com/loomio/loomio/pull/2970#discussion_r53220711>:
>
> > @@ -97,12 +97,7 @@ def viewed!(age_of_last_read_item = nil)
> >        reset_counts!
> >      end
> >
> > -    publish!
> > -  end
> > -
> > -  def publish!
> > -    publish_data = Simple::DiscussionSerializer.new(discussion, scope: { reader_cache: reader_cache })
> > -    MessageChannelService.publish(publish_data.as_json, to: user)
> > +    EventBus.broadcast('discussion_reader_viewed', discussion, self, user)
>
> I was trying to keep the broadcast super close to where the work is done.
>
> Right now there is problem that when you create a discussion, other tabs
> don't see that it is read.
>
> I don't think we're going to call DiscussionService#mark_as_read from
> DiscussionService#create are we?
>
> Seems like the tidiest thing to do is fire an event whenever a discussion
> is viewed (ie: discussion_reader.viewed!).
>
> There is heaps more I want to talk to you about today now I've had
> experience with this.. Would be great to pair again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/loomio/loomio/pull/2970/files#r53220711>.
>